File: languages/python/crypto_util.py
"""
Simple pure-Python cryptography utilities (wrapper around cryptography library).
This is a stub to be used by higher-level services. Requires `cryptography` package.
"""
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.primitives import serialization, hashes
import os, base64


def generate_rsa_keypair():
    priv = rsa.generate_private_key(public_exponent=65537, key_size=2048)
    pub = priv.public_key()
    return priv, pub


def hybrid_encrypt(plaintext: bytes, public_key) -> dict:
    # generate AES key
    aes_key = AESGCM.generate_key(bit_length=128)
    aesgcm = AESGCM(aes_key)
    iv = os.urandom(12)
    ct = aesgcm.encrypt(iv, plaintext, None)
    wrapped = public_key.encrypt(aes_key, padding.OAEP(mgf=padding.MGF1(algorithm=hashes.SHA256()), algorithm=hashes.SHA256(), label=None))
    return {"cipher": base64.b64encode(iv+ct).decode(), "wrappedKey": base64.b64encode(wrapped).decode()}


def hybrid_decrypt(cipher_b64: str, wrapped_b64: str, private_key) -> bytes:
    ivct = base64.b64decode(cipher_b64)
    iv = ivct[:12]
    ct = ivct[12:]
    wrapped = base64.b64decode(wrapped_b64)
    aes_key = private_key.decrypt(wrapped, padding.OAEP(mgf=padding.MGF1(algorithm=hashes.SHA256()), algorithm=hashes.SHA256(), label=None))
    aesgcm = AESGCM(aes_key)
    return aesgcm.decrypt(iv, ct, None)

File: languages/node/cryptoUtil.js
const crypto = require('crypto');

function generateRsaKeyPair() {
  return crypto.generateKeyPairSync('rsa', { modulusLength: 2048 });
}

function hybridEncrypt(plaintextBuffer, publicKeyPem) {
  const aesKey = crypto.randomBytes(16); // 128-bit
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-128-gcm', aesKey, iv);
  const ct = Buffer.concat([cipher.update(plaintextBuffer), cipher.final()]);
  const tag = cipher.getAuthTag();
  const wrapped = crypto.publicEncrypt({ key: publicKeyPem, padding: crypto.constants.RSA_PKCS1_OAEP_PADDING, oaepHash: 'sha256' }, aesKey);
  return { cipher: Buffer.concat([iv, ct, tag]).toString('base64'), wrappedKey: wrapped.toString('base64') };
}

function hybridDecrypt(cipherB64, wrappedB64, privateKeyPem) {
  const data = Buffer.from(cipherB64, 'base64');
  const iv = data.slice(0, 12);
  const tag = data.slice(data.length - 16);
  const ct = data.slice(12, data.length - 16);
  const wrapped = Buffer.from(wrappedB64, 'base64');
  const aesKey = crypto.privateDecrypt({ key: privateKeyPem, padding: crypto.constants.RSA_PKCS1_OAEP_PADDING, oaepHash: 'sha256' }, wrapped);
  const decipher = crypto.createDecipheriv('aes-128-gcm', aesKey, iv);
  decipher.setAuthTag(tag);
  const pt = Buffer.concat([decipher.update(ct), decipher.final()]);
  return pt;
}

module.exports = { generateRsaKeyPair, hybridEncrypt, hybridDecrypt };
